File: the-daily-starter/pages/api/daily-quote/AllQuotes.ts
import { NextApiRequest, NextApiResponse } from "next";
import { getAllDailyQuotes } from "@/lib/firebase/utils/firebaseService";
import { DailyQuote } from "@/lib/firebase/utils/firebaseInterfaces";

//api endpoint for fetching all daily quotes from firebase.
export default async function handler(req: NextApiRequest, res: NextApiResponse){

    if(req.method !== 'GET'){
        return res.status(405).json({ message: 'Method Not Allowed' });
    }

    try{
        const dailyQuotes: DailyQuote[] = await getAllDailyQuotes();
        return res.status(200).json(dailyQuotes);

    } catch (err) {
        console.error('An error occurred while fetching all quotes: ' + err);
        return res.status(500).json({ error: 'Failed to fetch all quotes' });
    }
}

File: backend/quoteservice/endpoint/allQuotes.go
package QuoteEndpoint

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"services/quoteservice"
	"time"
)

// handler to fetch all zen quotes from firebase database. Returns handler function.
func AllQuotesHandler(svc *quoteservice.Service) http.HandlerFunc {
	//actual function handler
	return func(w http.ResponseWriter, r *http.Request) {
		//http method has to be get
		if r.Method != http.MethodGet {
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
			return
		}

		//set timeout for getting context
		ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
		defer cancel()

		//call helper function from dailyservice
		quotes, err := svc.GetAllDailyQuotes(ctx)
		if err != nil {
			log.Printf("error fetching all quotee: %v", err)
			http.Error(w, "error fetching all quotes", http.StatusInternalServerError)
			return
		}

		//Encode the result into json
		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(quotes); err != nil {
			log.Printf("Error encoding response: %v", err)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			return
		}

	}
}
